debian/nxconvert.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH NXCONVERT 1 "June 2010"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
nxconvert \- convert a NeXus file between different on disk file formats
.SH SYNOPSIS
.B nxconvert
[-x|-h4|-h5|-d|-o keepws|-o table] [\fIinfile\fP [\fIoutfile\fP]]
.SH DESCRIPTION
NeXus supports different file formats for physical storage on disk or other media.
.B nxconvert
allows the user to convert a file to a different backend.
.PP
If no output file is supplied on the command line, the program asks for one 
interactively. Same if the input file is not given.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics,
.\" respectively.
.SH OPTIONS
The following options are supported:
.TP
.B -x 
make outfile XML 
.TP
.B -h4 
make outfile HDF4 (this is the default)
.TP
.B -h5 
make outfile HDF5
.TP
.B -d
make a XML based definition file used for validating NeXus files (i.e. remove the data).
.TP
.B -o keepws
the XML file created should preserve whitespace.
.TP
.B -o table
the XML file created should write the data in a table format where the columns and rows are easily imported into spreadsheet programs.
.SH SEE ALSO
.BR http://www.nexusformat.org
.br
.SH AUTHOR
nxconvert was originally written by Freddie Akeroyd and Ray Osborn.
.PP
This manual page was written by Tobias Richter 
.nh
<Tobias.Richter@diamond.ac.uk>
.hy
and may be used by others.


NEWS
Version 4.3.0
=============

If creating an external link you are no longer supposed to create the group 
beforehand. As HDF5 can handle groups natively, that is counterproductive 
and we would have to remove the empty group again (hoping that it is actually 
empty that is). It is easier to create the group in the API if required.

Version 4.2.0
=============

For the latest information see 

    http://www.nexusformat.org/Nexus_42_Release_Notes

==System Requirements==
'''MXML XML Parsing Library'''

Version 2.2.2 or higher of mxml is required. Earlier versions have a bug and the XML API will not work. This package can be downloded in [http://www.easysw.com/~mike/mxml/software.php both source and binary rpm form] and is also available as part of [http://fedoraproject.org/wiki/Extras/UsingExtras Fedora Extras]. IMPORTANT NOTE: Debian also provides the mxml package, but it based on 2.0 and will not work properly.

'''Python Interface'''
You will need both the numpy and ctypes modules to be available.  These are provided in both the Fedora and EPEL repositories.

==Building Notes==
===NAG F90/F95 Compiler===
The NAG compiler needs the '''-mismatch''' flag to be specified or else it will not compile NXmodule.f90 This is achieve by running configure with the '''FCFLAGS''' environment variable set to contain the flag e.g.
<pre>
env FCFLAGS="-mismatch" ./configure --with-f90=f95
</pre>
===HDF4 on Intel Macs===
There is a problem with the include file, hdfi.h (normally in /usr/local/include).  See http://coastwatch.noaa.gov/helparc/software/msg00069.html for details of the modifications necessary to fix it.

==New Features==
===C++ Interface===
See the [http://download.nexusformat.org/doxygen/html/classNeXus_1_1File.html doxygen documentation] and
[http://svn.nexusformat.org/code/branches/4.2/test/napi_test_cpp.cxx NeXus API test program]

===C++ Stream Like interface===
The idea is to provide an IOSteam like interface and allow you to type 
<pre>
    // create an entry and a data item
    File nf(fname, NXACC_CREATE);
    nf << Group("entry1", "NXentry") << Data("dat1", w, "int_attr", 3);
    nf.close();

    File nf1(fname, NXACC_RDWR);
    // add a double_attr to an existing setup
    nf1 >> Group("entry1", "NXentry") >> Data("dat1") << Attr("double_attr", 6.0);
    nf1.close();

    // read back data items
    File nf2(fname, NXACC_READ);
    nf2 >> Group("entry1", "NXentry") >> Data("dat1", w1, "int_attr", i, "double_attr", d);
    // alternative way to read d1
    nf2 >> Data("dat1") >> Attr("double_attr", d1);
</pre>
See also the [http://svn.nexusformat.org/code/branches/4.2/test/napi_test_cpp.cxx NeXus API test program]

===IDL Interface===
There is a new interface to RSI's Interactive Data Language, IDL for NeXus. This 
interface has to be considered beta. Nevertheless it is working most of the time. 
Known issues include:
* Compressed reading and writing do not work for HDF-4 files, probably because of a library version conflict on libz. 
* There is an issue using NXgetslab on 64 bit operating systems; expect a fix for this pretty soon.

===Python Interface===
There is now, thanks to Paul Kienzle, a supported interface for the python scripting language. Arrays are stored in numpy arrays and thus allow for efficient data manipulations.

==Changed Features==

==Known Issues==
See the comments on the IDL interface.

==Miscellaneous bug fixes==
The following items are bugs reported in previous releases and resolved in
the 4.2 release.

==Upcoming Features==

Version 4.1.0
=============

For the latest information see 

    http://www.nexusformat.org/Nexus_41_Release_Notes

==System Requirements==
'''MXML XML Parsing Library'''

Version 2.2.2 or higher of mxml is required. Earlier versions have a bug and the XML API will not work. This package can be downloded in [http://www.easysw.com/~mike/mxml/software.php both source and binary rpm form] and is also available as part of [http://fedoraproject.org/wiki/Extras/UsingExtras Fedora Extras]. IMPORTANT NOTE: Debian also provides the mxml package, but it based on 2.0 and will not work properly.

==Building Notes==
===NAG F90/F95 Compiler===
The NAG compiler needs the '''-mismatch''' flag to be specified or else it will not compile NXmodule.f90 This is achieve by running configure with the '''FCFLAGS''' environment variable set to contain the flag e.g.
<pre>
env FCFLAGS="-mismatch" ./configure --with-f90=f95
</pre>
===HDF4 on Intel Macs===
There is a problem with the include file, hdfi.h (normally in /usr/local/include).  See http://coastwatch.noaa.gov/helparc/software/msg00069.html for details of the modifications necessary to fix it.

==New Features==
* New types NX_INT64 and NX_UINT64 to suppport 64 bit integers (only available in HDF5 and XML) [http://trac.nexusformat.org/code/ticket/87 details].
* Python bindings are now included in the Windows install kit [http://trac.nexusformat.org/code/ticket/86 details]

==Miscellaneous bug fixes==
The following items are bugs reported in previous releases and resolved in
the 4.1 release.
* The Fortran 90 part of testsuite failed with the Absoft compiler on MacOSX (it passed with g95 and gfortran (4.2)) [http://trac.nexusformat.org/code/ticket/68 details here]
* NXputattr assumed NULL termination of NX_CHAR attributes, which is usually the case in C but not true for JAVA. A workaround is to add '\0' manually [http://trac.nexusformat.org/code/ticket/83 bug report]
* pkgconfig issue [http://trac.nexusformat.org/code/ticket/84 bug report]
* Build issue with MXML-2.3 [http://trac.nexusformat.org/code/ticket/91 bug report]
* XML buffer resizing performance issue [http://trac.nexusformat.org/code/ticket/92 bug report]
* Documentation is now installed to "datadir" (/usr/share) [http://trac.nexusformat.org/code/ticket/93 bug report]

Version 4.0.0
=============

For the latest information see 

    http://www.nexusformat.org/Nexus_4_Release_Notes

==System Requirements==
'''MXML XML Parsing Library'''


applications/NXconvert/nxconvert.cpp
/*-----------------------------------------------------------------------------
  NeXus - Neutron & X-ray Common Data Format
   
  Utility to convert a NeXus file into HDF4/HDF5/XML/...
 
  Author: Freddie Akeroyd, Ray Osborn
 
  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2 of the License, or (at your option) any later version.
 
  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.
 
  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 
  For further information, see <http://www.nexusformat.org>
 
 $Id$
-----------------------------------------------------------------------------*/

#include <iostream>
#include <stdexcept>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <vector>
#ifdef _MSC_VER
#else
#include <unistd.h>
#endif
#include "napi.h"
#include "nxconvert_common.h"
#include "tclap/CmdLine.h"

using std::runtime_error;
using std::string;
using std::vector;
using namespace TCLAP;

static const string NXCONVERT_VERSION("1.0");
static const string EMPTY("");

static void print_usage()
{
    printf("Usage: nxconvert [ -x | -h4 | -h5 | -d | -Dfile | -o keepws | -o table ] [ infile ] [ outfile ]\n");
}

int main(int argc, char *argv[])
{
  try {
    // set up the command line  arguments
    CmdLine cmd("Convert a NeXus file between different on disk file formats.",
		' ', NXCONVERT_VERSION, false);
    SwitchArg xmlArg("x", "xml", "Output file in xml format", false);
    vector<int> allowedHdf;
    allowedHdf.push_back(4);
    allowedHdf.push_back(5);
    ValuesConstraint<int> allowedHdfC(allowedHdf);
    ValueArg<int> hdfArg("h", "hdf", "Specify HDF version to write", false, 0, &allowedHdfC);
    SwitchArg defArg("d", "dfn",
		     "Output definition file used for validating NeXus files",
		     false);
    ValueArg<string> defValueArg("D", "definition",
		      "Output definition file used for validating NeXus files. Requires a definition name.",
			    false, EMPTY, "defintion");
    vector<string> allowedXml;
    allowedXml.push_back("keepws");
    allowedXml.push_back("table");
    ValuesConstraint<string> allowedXmlC(allowedXml);
    ValueArg<string> xmlSpecialArg("o", "outputxml",
				   "Special arguments for xml. keepws defines that the whitespace should be preserved. table specifies a format that is more easiliy imported into spreadsheet programs. Either option forces xml output",
				   false, "", &allowedXmlC);
    vector<Arg*> outputformats;
    outputformats.push_back(&hdfArg);
    outputformats.push_back(&xmlArg);
    outputformats.push_back(&defArg);
    outputformats.push_back(&defValueArg);
    outputformats.push_back(&xmlSpecialArg);
    cmd.xorAdd(outputformats);

    UnlabeledMultiArg<string> FileArgs("Files", "Name of input and output files.",
					false, EMPTY);
    cmd.add(FileArgs);
    // parse the arguments and configure converting the file
    cmd.parse(argc, argv);
    int nx_format = -1; // output format
    int nx_write_access = 0; // output write access
    int nx_read_access = NXACC_READ; // input read access
    string definition_name; // NEXUS_SCHEMA_BASE is NULL


    if (xmlArg.isSet()) {
      nx_format = NX_XML;
      nx_write_access |= NXACC_CREATEXML;
    }
    if (defArg.isSet() || defValueArg.isSet()) {
      nx_format = NX_DEFINITION;
      nx_write_access |= NXACC_CREATEXML;
      if (defValueArg.isSet())
	definition_name = defValueArg.getValue();
    }
    if (hdfArg.isSet()) {
      int hdf_type = hdfArg.getValue();
      if (hdf_type == 4) {
          nx_format = NX_HDF4;
          nx_write_access |= NXACC_CREATE4;
      } else if (hdf_type == 5) {
          nx_format = NX_HDF5;
          nx_write_access |= NXACC_CREATE5;
      }
    }
    if (xmlSpecialArg.isSet()) {
      nx_format |= NX_XML;
      nx_write_access |= NXACC_CREATEXML;
      string type = xmlSpecialArg.getValue();
      if (type.compare("keepws")) {
	nx_write_access |= NXACC_NOSTRIP;
	nx_read_access |= NXACC_NOSTRIP;
      }
      else if (type.compare("table")) {
	nx_write_access |= NXACC_TABLE;
      }
    }
    vector<string> file_args = FileArgs.getValue();
    string inFile, outFile;
    if (file_args.size() > 0)
    {
        inFile = file_args[0];
    }
    if (file_args.size() > 1)
    {
        outFile = file_args[1];
    }

    // do the actual conversion
    std::cout << "Converting " << inFile << " to " << nx_formats[nx_format]
	      << " NeXus file " << outFile << std::endl;
    
    if (convert_file(nx_format, inFile.c_str(), nx_read_access, outFile.c_str(), nx_write_access, definition_name.c_str()) != NX_OK) {
      std::cerr << "Conversion failed" << std::endl;
      return 1;
    }

  }
  catch (ArgException &e) {
    std::cerr << "PARSE ERROR:" << e.error() << " for arg " << e.argId()
	      << std::endl;
    return -1;
  }
  catch (runtime_error &e) {
    std::cerr << "RUNTIME ERROR:" << e.what() << std::endl;
    return -1;
  }

  // tell the user that everything went ok
  std::cout << "Convertion successful." << std::endl;
  return 0;
}

